### PR TITLE
Make bossbar use `EntityLayer` for visibility

### PR DESCRIPTION
**Pull request by Bafbi**
# Objective

We want things to be easy for the user as possible, so I used the same api principes from EntityLayer.

# Solution

`BossBarBundle` also have an `EntityLayerId` that says in what layer they are, and client see bossbar that are in his `VisibleEntityLayers`.

### CODE REVIEW

File: crates/valence_boss_bar/src/lib.rs [hunk 1]
```diff
@@ -96,148 +39,96 @@ impl Plugin for BossBarPlugin {
...
-        for &added_viewer in current_viewers.difference(old_viewers) {
-            if let Ok(mut client) = clients.get_mut(added_viewer) {
+    for (mut client, visible_entity_layers, old_visible_entity_layers) in clients_query.iter_mut() {
+        let old_layers = old_visible_entity_layers.get();
+        let current_layers = &visible_entity_layers.0;
+
+        for &added_layer in current_layers.difference(old_layers) {
+            if let Some((id, title, health, style, flags, _)) = boss_bars_query
+                .iter()
+                .find(|(_, _, _, _, _, layer_id)| layer_id.0 == added_layer)
```
[the remark this answers is not in the document]
Comment (Bafbi) [reply crates/valence_boss_bar/src/lib.rs:96]: I did think of that, but unfortunately that would not permit of having multiple bossbar per layer.
I also thought of creating a bossbar layer, but the way `valence_layer` are design do not really make it easy. I was trying to make bvh more generic and created a BlockPos Aabb implementation but I couldn't figure out why it wouldn't pass the test so I abandoned this idea.

File: crates/valence_boss_bar/src/lib.rs [hunk 2]
```diff
@@ -96,148 +39,96 @@ impl Plugin for BossBarPlugin {
...
+    for (id, part, entity_layer_id, pos) in boss_bars_query.iter() {
+        if let Ok(mut entity_layer) = entity_layers_query.get_mut(entity_layer_id.0) {
+            let packet = BossBarS2c {
+                id: id.0,
+                action: part.to_packet_action(),
+            };
+            if let Some(pos) = pos {
+                entity_layer
+                    .view_writer(pos.to_chunk_pos())
+                    .write_packet(&packet);
```
[the remark this answers is not in the document]
Comment (Bafbi) [reply crates/valence_boss_bar/src/lib.rs:67]: Yes, I frustrate myself after trying layer impl, and wanted the thing done and I forget to continue this logic.

File: crates/valence_packet/src/packets/play/boss_bar_s2c.rs [hunk 3]
```diff
@@ -4,29 +4,29 @@ use super::*;
...
-        title: Cow<'a, Text>,
+        title: Text,
         health: f32,
         color: BossBarColor,
         division: BossBarDivision,
         flags: BossBarFlags,
     },
     Remove,
     UpdateHealth(f32),
-    UpdateTitle(Cow<'a, Text>),
```
[the remark this answers is not in the document]
Comment (Bafbi) [reply crates/valence_packet/src/packets/play/boss_bar_s2c.rs:23]: Because I could'nt find how to update the title from the message in ChatMessageEvent, I'm really bad at dealing with the borrow checker.

File: crates/valence_boss_bar/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dyc3) [reply crates/valence_boss_bar/src/lib.rs:96]: Can you have multiple boss bars in vanilla? I can only recall seeing one at a time.

[the remark this answers is not in the document]
Comment (rj00a) [reply crates/valence_boss_bar/src/lib.rs:96]: You can. they'll be stacked on top of each other.

File: crates/valence_boss_bar/src/components.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    }
+}
+
+/// The style of a boss bar. This includes the color and division of the boss
+/// bar.
+#[derive(Component, Default)]
+pub struct BossBarStyle {
+    pub color: BossBarColor,
+    pub division: BossBarDivision,
+}
```
[the remark this answers is not in the document]
Comment (Bafbi) [reply crates/valence_boss_bar/src/components.rs:47]: would not be possible with generic system

File: crates/valence_packet/src/packets/play/boss_bar_s2c.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bafbi) [reply crates/valence_packet/src/packets/play/boss_bar_s2c.rs:23]: And maybe I don't understand `Cow` but will it not create a copy anyway when we encode it ?

File: crates/valence_protocol/src/packets/play/update_selected_slot_c2s.rs [hunk 5]
```diff
@@ -3,5 +3,5 @@ use super::*;
 #[derive(Copy, Clone, Debug, Encode, Decode, Packet)]
 #[packet(id = packet_id::UPDATE_SELECTED_SLOT_C2S)]
 pub struct UpdateSelectedSlotC2s {
-    pub slot: u8,
+    pub slot: u16,
```
Comment (dyc3) [inline crates/valence_protocol/src/packets/play/update_selected_slot_c2s.rs:6]: This change seems unrelated to this PR. If it's not relevant to boss bars, revert it.

File: crates/valence_inventory/src/lib.rs [hunk 6]
```diff
@@ -1235,11 +1235,11 @@ fn handle_update_selected_slot(
...
                     continue;
                 }
-                held.held_item_slot = convert_hotbar_slot_id(pkt.slot as u16);
+                held.held_item_slot = convert_hotbar_slot_id(pkt.slot);
 
                 events.send(UpdateSelectedSlotEvent {
                     client: packet.client,
-                    slot: pkt.slot,
+                    slot: pkt.slot as u8,
                 });
```
Comment (dyc3) [inline crates/valence_inventory/src/lib.rs:1243]: These changes seem unrelated to this PR. If it's not relevant to boss bars, revert them.

File: examples/boss_bar.rs [hunk 7]
```diff
@@ -43,23 +48,37 @@ fn setup(
...
+        CustomBossBar,
+    ));
+
+    commands.spawn((
+        CowEntityBundle {
+            position: Position::new([0.0, SPAWN_Y as f64 + 1.0, 0.0]),
+            layer: EntityLayerId(layer_id),
+            ..Default::default()
+        },
+        BossBarTitle("BigCow".color(NamedColor::Red)),
```
Comment (dyc3) [inline examples/boss_bar.rs:69]: nit
```suggestion
        BossBarTitle("Big Cow".color(NamedColor::Red)),
```

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: crates/valence_protocol/src/packets/play/update_selected_slot_c2s.rs (hunk 5 above)
Comment (Bafbi) [reply crates/valence_protocol/src/packets/play/update_selected_slot_c2s.rs:6]: Sure I can but it will just bloat the git history for nothing, pr are not shown, so what will be show is that this line has been modify, then later reversed to be rereversed just after. not sure to be really worth it.

Comment (dyc3) [reply crates/valence_protocol/src/packets/play/update_selected_slot_c2s.rs:6]: PRs are squashed merged, so each PR becomes one commit. If we need to revert one, it would be better to not have collateral damage. It'll also make git bisects a bit better.

**Review by dyc3 — APPROVED** [review]
